### ISSUE TITLE
[DevEx RFC] Marketplace Contract for TokenV2

### ISSUE THREAD

**Issue by davidiw** [aptos-labs/aptos-core#7798]
As we migrate into TokenV2, application contracts will need to be adjusted since TokenV1 used store and TokenV2 uses objects. In effect, this issue should encompass both CoinV1 and FungibleAsset to/from Object "swap".

With the introduction to Objects, we can determine if we would rather leverage objects to construct this or continue to use tables. In the interest of doing something fun and interesting, we will NOT use tables, but we will describe both solutions for the sake of `research`.

### Table Solution

Because objects require either a `TransferRef` or a `signer` to change ownership. This structure will need to be stored within a resource account. It could be shared by many participants, a marketplace, or just per user.

Each account stores the following struct:
```
struct Swap has key {
   original_owner: address,
    inner: Table<Object<ObjectCore>, SwapInfo>,
    // some events
}
```

where `SwapInfo` contains the logic critical for transfers:
```
struct SwapInfo has store {
    expected_type: String,
    amount: u64, // if 1, can handle on object path, otherwise must take fungible asset or coin path
}
```

To support adding objects to the store:
```
// set amount to 1 and expected_type to address of to
public fun swap_for_object<F, T>(account: &signer, from: Object<F>, to: Object<T>);

// set amount to amount and expected type to address of metadata
public fun swap_for_asset<F>(account: &signer, from: Object<F>, to: Object<Metadata>, amount: u64);

// set amount to amount and expected type to type info of coin
public fun swap_for_coins<F, T>(account: &signer, from: Object<F>, to: Coin<T>, amount: u64);
```

For the owner to extract:
```
public fun extract<T>(account: &signer, object: Object<T>);
```

To support the various different transfer options:
```
public fun swap_objects<F, T>(account: &signer, from: Object<F>, to: Object<T>);

public fun swap_assets<T>(account: &signer, from: Object<Metadata>, amount: u64, to: Object<T>);

public fun swap_coin<F, T>(account: &signer, from: Coin<F>, amount: u64, to: Object<T>);
```

Note: We can put either a CoinV1 or a FungibleAsset into a distinct object for storing in this swap.

Okay that all seems sensible, but there are a few caveats:
1. Having a resource account sucks, it is yet another slot
2. Tables are hard to parse
3. We don't really get the benefit of objects

### Pure Objects (Pure Love?)

Let's not use resource accounts and instead let's use objects owning objects, this is so much more fun...

Now we need to make a `SwapableObject`

```
struct SwapableObject has key {
    original_owner: address,
    owned_object: Object<ObjectCore>,
    expected_type: String,
    amount: u64, // if 1, can handle on object path, otherwise must take fungible asset or coin path
    // some events
}
```

Now when I want to put out an object for a swap, we can use all of the existing functions from above, but we instead create the above as an object and transfer ownership of our object to this object. Hazzuh, complete.

Wow... something seems a little to easy here... I guess the devil is in the details.

## Part 2: Add Indexer Support

Need support from @rtso (w/ @bowenyang007 and @larry-aptos as backup)

## Part 3: SDKs

Extend the Typescript and/or Python SDK with appropriate support to call into these functions. Note, you can do this in parallel of indexer support, but you'll need to read transaction writesets to get the SwapableObject's address so you can write end-to-end examples.

## Part 4: Write a Tutorial

Now that we have all the pieces, let's make sure we have a decent tutorial that introduces these new frameworks. We may also want to consider adding this as an AIP, if we find this reasonable as a standard.

**Comment by gregnazario** [comment]
Here is the PR for a marketplace that handles Coin based swaps, auctions, collection offers etc.

It would likely need to be added to for support for Fungible assets (both listing and payment), token swaps.
https://github.com/aptos-labs/aptos-core/pull/8112
